src/group_theory/monoid_localization.lean

```diff
@@ -406,8 +439,354 @@ lemma map_map {A} [comm_monoid A] {V} {g : Z →* A}
+      mapping `x` to an invertible element of `Z`, the homomorphism induced from the localization
+      of `X` at the submonoid generated by `x` to `Z`. -/
+  @[elab_with_expected_type]
+  noncomputable def away.lift {x} (hfx : is_unit (f x)) : away x →* Z :=
+  lift' f (λ s, classical.some hfx ^ classical.some s.2) $ λ s,
+  by rw [units.coe_pow, ←classical.some_spec hfx,
+         ←f.map_pow, classical.some_spec s.2]; refl
+
+  @[simp] lemma away.lift_of {x} (hfx : is_unit (f x)) (a) :
+    away.lift f hfx (of (submonoid.powers x) a) = f a :=
+  lift'_of _ _
+
+  @[simp] lemma away.lift_comp_of {x} (hfx : is_unit (f x)) :
+    (away.lift f hfx).comp (of $ submonoid.powers x) = f :=
+  lift'_comp_of _
+
+  /-- Given a `comm_monoid` `X` and elements `x y : X`, the homomorphism from the localization of
+      `X` at the submonoid generated by `x` to the localization of `X` at the submonoid generated
+      by `x * y` induced by the map sending `a : X` to the equivalence class of (a, 1) in
+      `away (x * y)`. -/
+  noncomputable def away_to_away_right (x y : X) : away x →* away (x * y) :=
+  away.lift (of $ submonoid.powers (x * y)) $
+  is_unit_of_mul_one _
+    (((of $ submonoid.powers (x * y)) : X → away (x * y)) y * away.inv_self (x * y)) $ by
+      rw [away.inv_self, ←mul_assoc, ←(of _).map_mul, ←mk_self (show _ ∈ submonoid.powers (x * y),
+            from ⟨1, pow_one _⟩), mk_eq_mul_mk_one (x * y) _]
+
+end away
+
+end monoid_localization
+
+namespace add_monoid_localization
+
+variables {A : Type*} [add_comm_monoid A] (x : A)
+
+/-- The localization of an `add_comm_monoid` `A` at the submonoid generated by an element
+    `x : A`. -/
+@[reducible] def away := add_monoid_localization A (add_submonoid.multiples x)
+
+attribute [to_additive add_monoid_localization.away] monoid_localization.away
+
+/-- Given an `add_comm_monoid` `A` and `x : A`, the inverse of `⟦(x, 0)⟧` in the localization of
+    `A` at the `add_submonoid` generated by `x`. -/
+def away.neg_self : away x := mk 0 ⟨x, 1, add_monoid.one_smul x⟩
+
+attribute [to_additive add_monoid_localization.away.neg_self] monoid_localization.away.inv_self
+
+end add_monoid_localization
+
+variables (X)
+
+/-- Given a `comm_monoid` `X`, the congruence relation defining the Grothendieck group of `X`:
+    the localization of `X` at itself. -/
+@[to_additive add_completion.r "Given an `add_comm_monoid` `X`, the additive congruence relation defining the Grothendieck group of `X`: the localization of `X` at itself."]
+protected def completion.r : con (X × X) :=
+⟨λ x y, (⊤ : submonoid X).r (x.1, ⟨x.2, trivial⟩) (y.1, ⟨y.2, trivial⟩),
+  ⟨λ _, con.refl _ _, λ _ _, con.symm _, λ _ _ _, con.trans _⟩, λ _ _ _ _, con.mul _⟩
+
+/-- The Grothendieck group of a `comm_monoid` `X`: the localization of `X` at itself. -/
+@[reducible, to_additive add_completion "The Grothendieck group of an `add_comm_monoid` `X`: the localization of `X` at itself."]
+def completion := (completion.r X).quotient
```

Comment:
It seems to me that your choice to define `completion` from first principles leads to a lot of duplication of code. I would probably have made `completion` an `abbreviation` for `monoid_localization X X` and called it a day...
The end user will usually know that it is an example of a localization anyway, and it immediately gives them access to the entire API for localizations (which will probably grow in the future).
So, I know that you chose this route for reasons of clarity, but I'm not sure if things actually get that much clearer...

Reply:
I can't do ```monoid_localization X X```; it would have to be ```monoid_localization x (⊤ : submonoid X)```, and to represent ```x : X``` as an element of ```⊤``` I have to write ```(⟨x, submonoid.mem_top x⟩ : (⊤ : submonoid X))```. I can ```to_additive``` all of the ```completion``` stuff. It just seemed to be enough of a 'nice' special case of a localization to merit making the most of the niceness.

Reply:
Ah, so would a little helper function from `X` to `⊤` help here?